synapse/storage/databases/main/event_federation.py

```diff
@@ -54,46 +54,195 @@ def __init__(self, database: DatabasePool, db_conn, hs):
+            clause, args = make_in_list_sql_clause(
+                txn.database_engine, "event_id", batch
+            )
+            txn.execute(sql % (clause,), args)
+
+            for event_id, chain_id, sequence_number in txn:
+                seen_events.add(event_id)
+                event_chains[chain_id] = max(
+                    sequence_number, event_chains.get(chain_id, 0)
+                )
+
+        # Check that we actually have a chain ID for all the events.
+        events_missing_chain_info = initial_events.difference(seen_events)
+        if events_missing_chain_info:
+            # This can happen due to e.g. downgrade/upgrade of the server. We
+            # raise an exception and fall back to the previous algorithm.
+            logger.info(
+                "Unexpectedly found that events don't have chain IDs in room %s: %s",
+                room_id,
+                events_missing_chain_info,
+            )
+            raise _NoChainCoverIndex(room_id)
+
+        # Now we look up all links for the chains we have, adding chains that
+        # are reachable from any event.
+        sql = """
+            SELECT
+                origin_chain_id, origin_sequence_number,
+                target_chain_id, target_sequence_number
+            FROM event_auth_chain_links
+            WHERE %s
+        """
+
+        # A map from chain ID to max sequence number *reachable* from any event ID.
+        chains = dict(event_chains)
+
+        # We need to take a copy of `event_chains` as we need to separate chain
+        # IDs / seq nos from the given events vs. reachable ones.
+        for batch in batch_iter(event_chains, 1000):
+            clause, args = make_in_list_sql_clause(
+                txn.database_engine, "origin_chain_id", batch
+            )
+            txn.execute(sql % (clause,), args)
+
+            for (
+                origin_chain_id,
+                origin_sequence_number,
+                target_chain_id,
+                target_sequence_number,
+            ) in txn:
+                if origin_sequence_number <= chains.get(origin_chain_id, 0):
+                    chains[target_chain_id] = max(
+                        target_sequence_number,
+                        chains.get(target_chain_id, 0),
+                    )
+
+                    # The chain ID / seq no of a given event is reachable from
+                    # a different event, discard that chain from the given events.
+                    if (
+                        target_chain_id in event_chains
+                        and target_sequence_number >= event_chains[target_chain_id]
+                    ):
+                        event_chains.pop(target_chain_id)
+
+        # Don't include the given events (since we're finding the auth chain of
+        # those events).
+        for chain_id in event_chains:
+            if event_chains[chain_id] == chains[chain_id]:
+                chains[chain_id] -= 1
```

Comment:
I'm finding the logic around this quite confusing, and I'm not sure I totally buy that it works. Though the tests do somewhat suggest it does.

Can we instead have `chains` start off being the empty set, so that it only contains *other* chains that are reachable from the initial events, and then add in the chains from `event_chains`? So something like:

```python
chains = set()

# Add all other chains reachable from initial set
for start, end in txn:
    if start.seq <= chains.get(start.id):
        chains[start.id] = start.seq

# Add initial set of chains, excluding the the seq corresponding to initial event
for id, seq in event_chains:
    chains[id] = max(seq - 1, chains.get(id))
```

Reply:
Oh, interesting. Don't spend too long trying it again then, it may be that its just complex logic no matter how you slice it.

Reply:
I agree that what you have looks cleaner, it is possible I had other bugs that got fixed though!

(Note that this also only works because we don't use a recursive implementation.)

Reply:
I was able to make this change pretty easily, I think it cleans it up a bit!